### DISCUSSION TITLE
Make "break" in switch case clauses optional

### DISCUSSION THREAD

**Discussion by The-Futurist** [dotnet/csharplang#2707] [Language Ideas]
That's it really, make these optional, perhaps enforce consistency (if one case has a break they must all have) but just make them optional (i.e. the absence of a final break generates code identical to that generated if the break had been present).

My suggestion is based in the needless noise this adds to code.

**Comment by HaloFour** [comment]
I'd rather see a statement form of `switch` expressions that allow for multi-statement arms.  Relitigating the explicit design choices of the `switch` statement seems like a non-starter.  Implementing this would only lead to confusion.

**Comment by CyrusNajmabadi** [comment]
> My suggestion is based in the needless noise this adds to code.

The noise is intentional to avoid errors that happen at a reasonably common pace in languages that have implicit fallthrough.  I've personally had this save my butt quite a few times.  

Given that this was a very intentional language decision, and changing now would likely be a large negative for people who want this behavior, i can't see this going anywhere.

**Comment by The-Futurist** [comment]
> > My suggestion is based in the needless noise this adds to code.
> 
> The noise is intentional to avoid errors that happen at a reasonably common pace in languages that have implicit fallthrough. I've personally had this save my butt quite a few times.
> 

What error does this avoid? I think you didn't read what I wrote @CyrusNajmabadi. 

> Given that this was a very intentional language decision, and changing now would likely be a large negative for people who want this behavior, i can't see this going anywhere.

What negative could you possibly be referring to?

**Comment by CyrusNajmabadi** [comment]
>What negative could you possibly be referring to?

The negative that the break is not explicit in source.

**Comment by CyrusNajmabadi** [comment]
> What error does this avoid?

In traditional C-like langauges with implicit fallthrough, it's not uncommon for unintentional fallthrough to happen.  As such, explicit fallthrough, and explicit breaking was chosen for C#.  To make it very clear in the code what the *intention* was.  The goal is not to keep C#'s semantics.  The goal is to keep C#'s *explicitness*.  The explicitness itself *is* the feature, and it was very intentionally designed and implemented that way.

Removing the explicitness, but keeping semantics, isn't seen as a positive as it just means that the code becomes more confusing for readers.  It's especially confusing/misleading for users coming from the languages C# built from.   They will see C#'s switches *with* breaks, and your new switches *without* breaks, and they'll likely think teh latter have implicit fallthrough, like those other languages.  

This will likely lead to errors in coding because people are led astray by this strange deviance in syntax/behavior.

**Comment by The-Futurist** [comment]
@CyrusNajmabadi you don't credit developers with much intelligence. We deal with far more abstruse characteristics than this without any trouble. The break isnt good at all as it implies one can omit it, clearly you're prepared to tolerate that confusion! This is a no brainer, case doesn't support fall through period and you may add a break if you want to - there that wasn't hard was it?

**Comment by canton7** [comment]
> This is a no brainer, case doesn't support fall through period and you may add a break if you want to - there that wasn't hard was it?

As a complete outsider, can I recommend that you dial the attitude back a bit? Being constructive tends to make people want to cooperate with you a bit more.

> case doesn't support fall through period and you may add a break if you want to

Note that you *can* already omit the break and fallthrough, in one specific case:

```csharp
case A:
case B:
    // Stuff
    break;
```

Under your proposal, that becomes ambiguous. So your rule is actually "Case doesn't support fall-through period **unless** the case is empty and is above another case. Break is never required **unless** you want to avoid this fall-through." Suddenly life got a bit more complex.

Imagine that you, as a developer coming from another C-like language, see this:

```csharp
case A:
    Console.WriteLine("A");
case B:
case C:
    Console.WriteLine("C");
```

That's confusing, right? You would expect cases A and B to fall through into C. If you then learnt that C# *doesn't* support fall-through, you would then adjust your expectations and not expect B to fall through into C. You then get a nasty surprise when you realise that, actually, C# *does* fall through in this case. There's no way that you wouldn't come away with a nasty taste in your mouth.

> I'd rather see a statement form of switch expressions that allow for multi-statement arms

Agreed. It's doing to be much easier to invest in our new switch expression syntax, rather than butcher the legacy syntax.

**Comment by HaloFour** [comment]
@Korporal 

No, `case` does support fall-through, as is evidenced by the fact that the language allows this when there are no statements within a case arm.

**Comment by The-Futurist** [comment]
> @Korporal
> 
> No, `case` does support fall-through, as is evidenced by the fact that the language allows this when there are no statements within a case arm.

@HaloFour - In which case someone needs to get the [documentation ](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/switch) and the diagnostic message corrected:

`
Only one switch section in a switch statement executes. C# doesn't allow execution to continue from one switch section to the next. Because of this, the following code generates a compiler error, CS0163: "Control cannot fall through from one case label (<case label>) to another."`

**Comment by canton7** [comment]
@Korporal Read [the top of this section](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/switch#the-switch-section)

> Each switch section contains **one or more case labels** (either a case or default label) followed by one or more statements.

That document doesn't explicitly call it falling through, but it is clear that you're allowed to stack `case` labels on top of each other, which is what @HaloFour was referring to, and what I also referred to in my comment above.

The documentation is correct that execution cannot pass from one *switch section* to another, but you can have multiple case labels on a switch section.

**Comment by The-Futurist** [comment]
@canton7 - Learning a new language often includes being surprised, all I'm suggesting is we make this optional, a decorative keyword (it has no functionality) to be used IF someone so chooses.

The keyword `private `is optional on class members - one may be explicit if they wish or they may not so the concept I'm advocating is hardly radical here.
```cs

case A:
    Console.WriteLine("A");
case B:
case C:
    Console.WriteLine("C");
```

Err - no, that isn't confusing to me I know exactly what it does.

**Comment by The-Futurist** [comment]
> @Korporal Read a [little further down](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/switch#the-switch-section)
> 
> > Each switch section contains **one or more case labels** (either a case or default label) followed by one or more statements.
> 
> That document doesn't explicitly call it falling through, but it is clear that you're allowed to stack `case` labels on top of each other, which is what @HaloFour was referring to, and what I also referred to in my comment above.
> 
> The documentation is correct that execution cannot pass from one _switch section_ to another, but you can have multiple case labels on a switch section.

Why are you arguing and what are you arguing about? It doesn't matter what he may have meant, he called me out as being incorrect and I chose to rebut that.

**Comment by canton7** [comment]
> Learning a new language often includes being surprised

A good aim, though, is to decrease the number of surprises which might cause bugs.

> Err - no, that isn't confusing to me I know exactly what it does.

I specifically said:

> Imagine that you, as a developer coming from another C-like language, see this:

I didn't ask whether you, with all of the C# knowledge that you currently have, would find this confusing. I asked you to imagine that you were a new C# developer coming from another C-like language. I'm pretty confident in saying that most people would find the complex rule of "break statements are implicit *unless* two case labels are stacked on top of each other, in which case there is no implicit break statement" unnecessarily confusing. It's hardly the [pit of success](https://ericlippert.com/2007/08/14/c-and-the-pit-of-despair/).

We've got a brand new opportunity now, in the form of switch expressions (which, importantly, *don't* have break statements). It would IMO be a much better effort to get multi-statement arms on these.

**Comment by canton7** [comment]
> Why are you arguing and what are you arguing about? It doesn't matter what he may have meant, he called me out as being incorrect and I chose to rebut that.

Your point was "it can never be confusing to omit `break`, because there are no cases in C# where fallthrough is allowed". His point was, I believe, the same as mine -- fall-through *is* allowed in the case of multiple case statements stacked on top of each other, and this fact makes your proposal confusing. This is fairly clear from (emphasis mine):

> No, case does support fall-through, **as is evidenced by the fact that the language allows this when there are no statements within a case arm.**

Your message very clearly said that he was wrong, and cited the documentation as evidence. I pointed you to the bit of the document which said that, of course, he was right.

I think we all know that you can stack case statements on top of each other, so it came as a surprise when you appeared to claim that @HaloFour was wrong to state it.

**Comment by The-Futurist** [comment]
@canton7 - I will tell you that the choice to make explicit accessibility optional was a source of confusion for me very early on. Look at a simple console app - the `Main` method is `private` yet is obviously invoked externally.

When I first looked at a console app in 2003 or whenever it was I assumed default accessibility was `public` because I assumed `Main` must be `public` because it is invoked from outside the class.

All I'm saying here is the potential to confuse the newcomer is ever present, people learn, we're not dummies.

**Comment by canton7** [comment]
> When I first looked at a console app in 2003 or whenever it was I assumed default accessibility was public because I assumed Main must be public because it is invoked from outside the class.

I'd argue that that was a mistake in the language design, and access should always have been made explicit. It appears that most coding standards these days agree.

If there's confusion to newcomers, the aim should be to reduce it, not to add to it. See the "Pit of Success" link I posted above.

At least something like incorrect access isn't a very common source of bugs.

**Comment by canton7** [comment]
I'll say it again: switch statements are going to go the way of the dodo, especially if and when switch expressions get multi-expression arms. There's no point in trying to do anything to them - focus on switch expressions instead.

**Comment by HaloFour** [comment]
Yes, developers have to learn.  But when things are implicit there is an added cognitive burden, especially when the behavior is different from previous versions, different depending on scenarios, and different from various other languages that have similar syntax.  That cognitive burden increases the possibility of getting it wrong, and because it's implicit the compiler can do nothing to help the developer to avoid that problem.  Being explicit completely eliminates that ambiguity and the cost of being a little more verbose.  We write the code once, but we (and everyone else) has to read it significantly more often.  The ergonomics of the language should be optimized towards clarity, not eliminating verbosity.

**Comment by The-Futurist** [comment]
> > Why are you arguing and what are you arguing about? It doesn't matter what he may have meant, he called me out as being incorrect and I chose to rebut that.
> 
> Your point was "it can never be confusing to omit `break`, because there are no cases in C# where fallthrough is allowed". His point was, I believe, the same as mine -- fall-through _is_ allowed in the case of multiple case statements stacked on top of each other, and this fact makes your proposal confusing. This is fairly clear from (emphasis mine):
> 
> > No, case does support fall-through, **as is evidenced by the fact that the language allows this when there are no statements within a case arm.**
> 
> Your message very clearly said that he was wrong, and cited the documentation as evidence. I pointed you to the bit of the document which said that, of course, he was right.
> 
> I think we all know that you can stack case statements on top of each other, so it came as a surprise when you appeared to claim that @HaloFour was wrong to state it.

@canton7 - This argument rests entirely on your choice to apply the term _fall through_ to that scenario which the documentation does not do. In fact the documentation clearly states that a sequence of "empty" case labels like that constitute a single **switch section**, it does not in any way even imply that control is "falling through" - this is a term you and @HaloFour chose to apply but that choice does not make it a legitimate use.

This may seem rather pedantic but I was accused of being wrong and I rebutted that, I have that right surely?

**Comment by The-Futurist** [comment]
> > When I first looked at a console app in 2003 or whenever it was I assumed default accessibility was public because I assumed Main must be public because it is invoked from outside the class.
> 
> I'd argue that that was a mistake in the language design, and access should always have been made explicit. It appears that most coding standards these days agree.
> 
> If there's confusion to newcomers, the aim should be to reduce it, not to add to it. See the "Pit of Success" link I posted above.
> 
> At least something like incorrect access isn't a very common source of bugs.

Well we certainly agree here - I too think accessibility should have been explicit too.

**Comment by The-Futurist** [comment]
> I'll say it again: switch statements are going to go the way of the dodo, especially if and when switch expressions get multi-expression arms. There's no point in trying to do anything to them - focus on switch expressions instead.

@canton7 - Yes, that's a fair point I don't disagree with you, you do make a good point - but C# will always be backward compatible of course.

**Comment by canton7** [comment]
> @canton7 - This argument rests entirely on your choice to apply the term fall through to that scenario which the documentation does not do. In fact the documentation clearly states that a sequence of "empty" case labels like that constitute a single switch section, it does not in any way even imply that control is "falling through" - this is a term you and @HaloFour chose to apply but that choice does not make it a legitimate use.

@HaloFour was *very* clear that they used the term "fall-through" to include two case statements stacked on top of each other. Most other languages would term this "fall-through" as well. If you had a problem with the specific use of the *term* "fall-through" (rather than the wider point that was being made), it might have been more constructive to state that, rather than stating outright that that whole point being made was incorrect.

> Well we certainly agree here - I too think accessibility should have been explicit too.

Right, so you made an equivalence between the fact that accessibility can be implicit, and the proposal to have implicit break statements ("The keyword private is optional on class members - one may be explicit if they wish or they may not so the concept I'm advocating is hardly radical here"), but you also think that accessibility should be explicit? Surely that's an argument *against* allowing implicit break statements?

**Comment by The-Futurist** [comment]
> Yes, developers have to learn. But when things are implicit there is an added cognitive burden, especially when the behavior is different from previous versions, different depending on scenarios, and different from various other languages that have similar syntax. That cognitive burden increases the possibility of getting it wrong, and because it's implicit the compiler can do nothing to help the developer to avoid that problem. Being explicit completely eliminates that ambiguity and the cost of being a little more verbose. We write the code once, but we (and everyone else) has to read it significantly more often. The ergonomics of the language should be optimized towards clarity, not eliminating verbosity.

However removing the need for `break` does not mean the code isn't explicit. For example nobody (I assume) would argue the latter code block is not explicit and would offer improved readability with the `continue`

```cs
            for (int I = 0; I < 10; I++)
            {
                // do something
                continue;
            }

            for (int I=0; I < 10; I++)
            {
                // do something
            }
```

A closing brace is _itself_ explicit and it's behavior is very well defined - it is in a loop and it is in a `case` if the break were made optional.

**Comment by canton7** [comment]
Anyway, many people smarter than either of us considered this possibility for a long time, and decided on the current behaviour. I don't think there's any chance of changing that.

**Comment by canton7** [comment]
> A closing brace is itself explicit and it's behavior is very well defined - it is in a loop and it is in a case if the break were made optional.

Hmm? There's no closing brace in a case.

**Comment by The-Futurist** [comment]
@canton7 - The documnation does not use the term "fall through" to describe this scenario with `case: 2` and `case 3:`

```cs
      switch (caseSwitch)
      {
          case 1:
              Console.WriteLine("Case 1");
              break;
          case 2:
          case 3:
              Console.WriteLine($"Case {caseSwitch}");
              break;
          default:
              Console.WriteLine($"An unexpected value ({caseSwitch})");
              break;
      }
```
It describes it as a switch **section** and clearly states that a section may have _one or more_ labels.

Please stop attempting to correct me on this point, it's getting us nowhere, accept the fact that the term "fall through" is never used to describe the above, if you or @HaloFour perceive it as "falling through" then fine, but it isn't, I'm sorry but the documentation is the documentation.

**Comment by canton7** [comment]
I have never claimed that the documentation said otherwise. Both myself and @HaloFour  made it *very* clear that we were using the term "fall-through" to refer to stacking cases on top of each other. You have ignored the points being made, instead spending effort on objecting to this use of the term "fall-through" (in quite an obscure way). Please accept that we used the term "fall-through" to include "stacking cases on top of each", and re-read the points being made with this in mind.

**Comment by The-Futurist** [comment]
> > A closing brace is itself explicit and it's behavior is very well defined - it is in a loop and it is in a case if the break were made optional.
> 
> Hmm? There's no closing brace in a case.

@canton7 - Yes they are optional, I personally think they should be required to delineate the block but oddly the team didn't care, similarity to C was given a high priority, which is a shame.

```cs
            switch (args[0])
            {
                case "A":
                    {
                        ;
                    }
            }
```

I'm arguing now that the above is better than this:

```cs
            switch (args[0])
            {
                case "A":
                    ;
                    break;
            }
```

But this will never change now, its too late.

**Comment by canton7** [comment]
Well, braces in a case aren't really part of the case -- they're just a way of introducing a new scope, the same as if you introduced them in lots of other places. This is perfectly legal:

```csharp
case A:
    Console.WriteLine("First")
    {
        Console.WriteLine("Second");
    }
    Console.WriteLine("Third");
    break;
```

Given this, I'm afraid I'm unclear on what point you were trying to make with cases and closing braces.

------------------

> But this will never change now, its too late.

Right, so why raise it? Lots of things will never change.

**Comment by HaloFour** [comment]
For reference, here's what Java is shipping as a preview in JDK 13:

```java
// as a statement
switch (e) {
    case FOO -> System.out.println("FOO!");
    case BAR -> {
        System.out.println("Multiple statements!");
        System.out.println("BAR!");
    }
    default -> throw new IllegalArgumentException();
}

// as an expression
String result = switch (e) {
    case FOO -> "FOO!";
    case BAR -> {
        System.out.println("Multiple statements!");
        yield "BAR!";
    }
    default -> throw new IllegalArgumentException();
};
```

**Comment by canton7** [comment]
It's annoying when Java beats us to getting something cool...

**Comment by The-Futurist** [comment]
> Well, braces in a case aren't really part of the case -- they're just a way of introducing a new scope, the same as if you introduced them in lots of other places. This is perfectly legal:
> 
> ```cs
> case A:
>     Console.WriteLine("First")
>     {
>         Console.WriteLine("Second");
>     }
>     Console.WriteLine("Third");
>     break;
> ```
> 
> Given this, I'm afraid I'm unclear on what point you were trying to make with cases and closing braces.
> 
> > But this will never change now, its too late.
> 
> Right, so why raise it? Lots of things will never change.

I raised making break optional, not making braces compulsory. The former is not a breaking change, the latter would be.

**Comment by canton7** [comment]
I don't get why you raised braces at all. [This entire comment](https://github.com/dotnet/csharplang/issues/2707#issuecomment-517724627) was about braces, and I still don't understand the point you were trying to make.

**Comment by The-Futurist** [comment]
@canton7 - Really? I showed you a loop and how we do not enforce that we write continue at the end, the continue is a natural defined behavior for a loop when control reaches the closing brace. My point there was that just as we can omit continue in the loop and remain explicit we can do the same for a case.

The continue would be redundant the break however is enforced (yet functionally redundant).

**Comment by canton7** [comment]
Right - you talked about the closing brace of a loop. Case statements don't have closing braces. I don't see why an argument about closing braces in loops applies to case statements which don't have braces.

Case statements are much more akin to `goto` labels. There are no braces with `goto` either, and you always "fall across" a goto label.

**Comment by HaloFour** [comment]
There also isn't consensus as to what C# should do if `break` could be omitted: #603

**Comment by The-Futurist** [comment]
@HaloFour - Well that's a separate thing altogether, not sure I'm a fan...let me look...

**Comment by CyrusNajmabadi** [comment]
> This is a no brainer, case doesn't support fall through period and you may add a break if you want to - there that wasn't hard was it?

That would likely cause confusion. Which is something we wanted to avoid here.

**Comment by CyrusNajmabadi** [comment]
> @HaloFour - In which case someone needs to get the documentation and the diagnostic message 

That documentation is correct. It specifically discusses a switch *section*. One switch section cannot fall through to another, ever.   Within a single switch section though, you could have a *case label* fall through to the next case label (if it has no statements). 

Sections and case labels are different things

**Comment by CyrusNajmabadi** [comment]
>you don't credit developers with much intelligence. We deal with far more abstruse characteristics than this without any trouble.  

It's not a matter of intelligence.  It's a matter of avoiding common problems, mistakes and confusion.

**Comment by HaloFour** [comment]
@CyrusNajmabadi 

And that's technically correct (the best kind of correct), but that may not be how developers intuit the behavior of the feature.  Either way, I prefer that the explicitness that the compiler forces prevents said misinterpretation from resulting in bugs.

**Comment by CyrusNajmabadi** [comment]
>  Either way, I prefer that the explicitness that the compiler forces prevents said misinterpretation from resulting in bugs.

Me too! :-D

**Comment by theunrepentantgeek** [comment]
@Korporal, you wrote

> @CyrusNajmabadi you don't credit developers with much intelligence.

Most developers are quite intelligent.

They're also some combination of one or more of overly stressed, recently distracted, ravenously hungry, deadline pressured, under caffeinated, confused by the test results, toddler tired, regrettably hungover, time conscious, new to the language, recently promoted, intending to quit, justifiably angry, secretly delighted, clinically depressed, and under appreciated.

Any one of which can easily result in minor syntactic details being overlooked.

The explicitness of requiring each switch section to finish with **break**, **return** or **throw** helps to ensure that the most common failure modes result in compilation errors, not subtle bugs.

**Comment by The-Futurist** [comment]
@theunrepentantgeek - It's interesting to see that other languages like PL/I, Pascal, COBOL, Visual Basic, Eiffel, RPG and FORTRAN never supported "fall through" in any way, the designers clearly recognized there was no value in such a capability, it was an alien concept except in C.

The C language was the first to do this and it has contaminated later derivatives even C#.

Developers using non-C languages never had any problems because the concept of "fall through" was never even an option, confusion simply did not arise.

Basing a new language on one of the weakest high level languages carries a cost and C# is paying that cost.

**Comment by theunrepentantgeek** [comment]
> The C language was the first to do this and it has contaminated later derivatives even C#.
> Basing a new language on one of the weakest high level languages carries a cost and C# is paying that cost.

You're just about making my point for me.

Experienced developers see "C style syntax" and expect **case** fall-through unless the code explicitly says otherwise. This is true whether their experience is with C itself, or with one of the big three C-derived languages (C++, Java, C#).

The contamination *has already happened*. It can't be undone.

**Comment by CyrusNajmabadi** [comment]
> Basing a new language on one of the weakest high level languages carries a cost and C# is paying that cost.

And?  This is where we are.  We can't re-litigate the past.  The choices have been made with full understanding of hte details you've brought up and with a clear understanding of the user base and lineage of our language.

This is not PL/1#.  If you want that, you'd likely have a better chance just building it yourself.  This is C# and C# is going to continue it's development in this way for the foreseeable future.

Perhaps this just isn't the right language for you?  You might want to consider many of the others out there.  VB, as you mentioned, behaves in the manner you like here.  It has full access to the wealth of functionality on the .net platform if htat matters to you.  If you'd prefer a more functional language, F# might be right up your alley as well.
